File: benchmark/haystack_9193/repro_script.py
import os
from unittest.mock import patch

from haystack import Pipeline
from haystack.components.agents import Agent
from haystack.components.joiners import ListJoiner
from haystack.dataclasses import ChatMessage
from typing import List

# Import the actual components we will instantiate
from haystack.components.generators.chat import OpenAIChatGenerator
from haystack.components.websearch import SerperDevWebSearch
from haystack.tools import ComponentTool

# --- Setup ---
# 1. Set dummy API keys to satisfy the __init__ methods of the real components.
#    No network calls will be made, but the components expect these to be present.
os.environ["OPENAI_API_KEY"] = "DUMMY_KEY"
os.environ["SERPERDEV_API_KEY"] = "DUMMY_KEY"


# --- Verification Logic ---
# The try/except block is placed around the code that defines the pipeline,
# as the bug occurs during the pipeline.connect() call.
print("--- Attempting to connect to the 'agent.messages' output socket. ---")
print("--- This is expected to fail with a PipelineConnectError. ---")

try:
    # 1. Instantiate the REAL components. This is the key change. By creating
    #    real instances, they will pass all of Haystack's internal validation
    #    checks (like having a __haystack_input__ attribute) when they are
    #    passed to the Agent constructor.
    chat_generator = OpenAIChatGenerator()
    web_search_component = SerperDevWebSearch()

    # 2. Proceed with the setup from the user's report using the real components.
    #    The Agent will now initialize successfully because it receives a valid generator.
    search_tool = ComponentTool(component=web_search_component)
    agent = Agent(chat_generator=chat_generator, tools=[search_tool])

    pipeline = Pipeline()
    pipeline.add_component("agent", agent)
    pipeline.add_component("joiner", ListJoiner(List[ChatMessage]))

    # 3. This is the exact line that triggers the bug in the faulty version.
    #    The Agent class in this version does not have a 'messages' output socket,
    #    so trying to connect to it will raise a PipelineConnectError.
    pipeline.connect("agent.messages", "joiner.values")

    # If this line is reached, the bug was not reproduced.
    print("\n--- SCRIPT FINISHED UNEXPECTEDLY ---")
    print("FAILURE: The bug was NOT reproduced. The connection was made successfully.")
    exit(0)

except Exception as e:
    print(f"\nSUCCESS: The script failed with an error as expected.")
    print(f"Error Type: {type(e).__name__}")
    print(f"Error Message: {e}")

    # Verify that the error is the specific PipelineConnectError we expect.
    expected_error_text = "agent.messages does not exist"
    if "PipelineConnectError" in type(e).__name__ and expected_error_text in str(e):
        print("\nVerification successful: The error message matches the bug report.")
        # Exit with 1 to signal to the test runner that the bug was found.
        exit(1)
    else:
        print("\nVerification failed: The error did not match the expected bug.")
        exit(0)

File: benchmark/ai_4761/repro_script.js
import assert from 'node:assert';

// --- Test Setup ---

// 1. Simulate Vue's reactive data.
//    Vue 3 uses JavaScript's Proxy objects to make data reactive.
//    This is a simple representation of a reactive 'messages' array.
const reactiveMessages = new Proxy(
  [{ role: 'user', content: 'hello' }],
  {},
);

// 2. This function simulates the buggy part of the `useChat` hook.
//    It attempts to use `structuredClone` on the reactive data.
function simulateBuggyClone(messages) {
  console.log('--- Attempting to clone a reactive Vue Proxy object... ---');
  // This is the exact function call that causes the error.
  // The standard `structuredClone` cannot handle Proxy objects.
  return structuredClone(messages);
}


// --- Verification Logic ---
console.log('--- This test will now directly trigger the DataCloneError. ---');

try {
  // Call the function that contains the bug.
  simulateBuggyClone(reactiveMessages);

  // If this line is reached, it means structuredClone succeeded,
  // and the bug was not reproduced.
  console.log("\nFAILURE: The bug was NOT reproduced.");
  console.log("The structuredClone operation succeeded unexpectedly.");
  process.exit(0);

} catch (error) {
  // We expect an error. Now we verify it's the correct one.
  console.log(`\nSUCCESS: The script failed with an error as expected.`);
  console.log(`Error Type: ${error.name}`);
  console.log(`Error Message: ${error.message}`);

  // The bug is that a `DataCloneError` is thrown.
  const expectedErrorName = 'DataCloneError';

  if (error.name === expectedErrorName) {
    console.log("\nVerification successful: The error is a DataCloneError, which confirms the bug.");
    // Exit with 1 to signal to the test runner that the bug was found.
    process.exit(1);
  } else {
    console.log("\nVerification failed: The error was not the expected DataCloneError.");
    process.exit(0);
  }
}
